### DISCUSSION TITLE
Full fine-tune

### DISCUSSION THREAD

**Discussion by kopyl** [PixArt-alpha/PixArt-alpha#43] [Q&A]
Could you please post a brief tutorial on how to fully fine-tune this model on, say, the pokemon dataset?

Right now it's very confusing for me :(
More questions than answers.

But this model is very cool, i was not that excited even about SDXL.

https://huggingface.co/datasets/lambdalabs/pokemon-blip-captions

**Comment by kopyl** [comment]
Some of the questions:
1. How do I specify LR? (UPD: answer found: in a config file: `optimizer = dict(type='AdamW', lr=2e-5, weight_decay=3e-2, eps=1e-10)`)
2. How do I use HF dataset?

You said:
> You ONLY need to change the config file in [config](https://github.com/PixArt-alpha/PixArt-alpha/blob/master/configs/pixart_config) and dataloader in [dataset](https://github.com/PixArt-alpha/PixArt-alpha/blob/master/diffusion/data/datasets).

so

3. How do I need to change the config file?
4. How do I change the dataloader
5. What is `PixArt_xl2_img512_internalms.py`? Specifically what does `internalms` mean?
6. How do you decide which config to use? Only based on image size? Or are there any other criteria?

You said:
> The directory structure for SAM dataset is:
```
cd ./data

SA1B
├──images/
│  ├──sa_xxxxx.jpg
│  ├──sa_xxxxx.jpg
│  ├──......
├──partition/
│  ├──part0.txt
│  ├──part1.txt
│  ├──......
├──caption_feature_wmask/
│  ├──sa_xxxxx.npz
│  ├──sa_xxxxx.npz
│  ├──......
├──img_vae_feature/
│  ├──train_vae_256/
│  │  ├──noflip/
│  │  │  ├──sa_xxxxx.npy
│  │  │  ├──sa_xxxxx.npy
│  │  │  ├──......
```

so

7.  What is partition?
8. What is `caption_feature_wmask` and how do I get it?
9. What is `img_vae_feature` and how do I get it?
10. What is `train_vae_256` and how do I get it?
11. What is `noflip` and how do I get it?
12. What's the SAM dataset and where can I find it and find instructions on how to build it?

In my dataset I only have 1. Images; 2. Captions. Why would I need anything else for a simple text to image training?

13. How do you specify to save a training checkpoint each, say, 1k steps?
14. How do you convert a training checkpoint to a Diffusers model which you can load like `PixArtAlphaPipeline.from_pretrained("TUNED_MODEL_DIRECTORY", torch_dtype=torch.float16)`

**Reply by lawrence-cj** [reply]
If you don't want to look throught the whole training code in this repo, maybe using diffusers training pipeline instead is a better choice.
Here are some answers FYI:
3. you need to change the data related code in the specific config file, since if you may know the repo better, we change the dataset and dataloader here `data = dict(type='InternalDataMS', root='MJData', image_list_json=image_list_json, transform='default_train', load_vae_feat=True)` and `type is the dataset class`, `root is where the dataset is`, `image_list_json contains prompt, image height, image width and other image related information`.
4. you need to change the dataset according to your data's structure, not every dataset has exactly the same format, so we [here ](https://github.com/PixArt-alpha/PixArt-alpha/blob/master/diffusion/data/datasets/InternalData_ms.py) have several dataset class for you to construct your own.
5. `PixArt_xl2_img512_internalms.py` is what we use for 512px training, if you look carefully, [Here ](https://github.com/PixArt-alpha/PixArt-alpha/blob/fe0cb78065d64c18ecd8955a04e4f29138d47946/configs/pixart_config/PixArt_xl2_img512_internalms.py#L9C4-L9C61) is the difference between 'ms' with no 'ms' suffix file. 'ms' stands for multi-scale image input during training.
6. You need to use different config file for different dataset, model, image-size, multi-scale training, also you can create your own config file, like you said, if you want to train on the [pokemon](https://huggingface.co/datasets/lambdalabs/pokemon-blip-captions) dataset.
7. We split the SAM dataset into several parts, that's why we have 'partition' here, the information in the txt file is only the image path like 'dir/sa_xxxxx.jpg'.
8. `caption_feature_wmask` is the text feature extracted by [T5_model](https://github.com/PixArt-alpha/PixArt-alpha/blob/fe0cb78065d64c18ecd8955a04e4f29138d47946/tools/extract_features.py#L21). 
9. `img_vae_feature` is similar but the vae feature from [vae model](https://github.com/PixArt-alpha/PixArt-alpha/blob/fe0cb78065d64c18ecd8955a04e4f29138d47946/tools/extract_features.py#L53).
10-11. `traini_vae_256 ` and `noflip` are just two folders, nevermind.
The reason why you need `img_vae_feature` and `caption_feature_wmask` is only for training faster, if you can prepare these two features in advance, these two processes will not appear in the whole training process. Otherwise, in one iteration, you need to first extract the text feature using T5 and img feature using VAE, which is time consuming and occupy a lot of GPU memory.
12. [Here ](https://github.com/PixArt-alpha/PixArt-alpha/blob/fe0cb78065d64c18ecd8955a04e4f29138d47946/scripts/train.py#L124) is how we save each customed step and [here ](https://github.com/PixArt-alpha/PixArt-alpha/blob/fe0cb78065d64c18ecd8955a04e4f29138d47946/configs/pixart_config/PixArt_xl2_img512_internalms.py#L30C10-L30C10) is how to set it.
13. [Here ](https://github.com/PixArt-alpha/PixArt-alpha/blob/master/tools/convert_pixart_alpha_to_diffusers.py) is how to convert the ckpt into diffusers format.
14. refer to Q3: 'image_list_json'

**Thank you for your patient to give these questions which may confusing others too, I hope you can look through the codebase in this repo, which I think is convinient for further using it. We will upload a toy dataset and json file for everyone to better understant the codebase.**

**Reply by kopyl** [reply]
@lawrence-cj thank you.

5.1 What is "multi-scale image input during training"?

**Reply by kopyl** [reply]
@lawrence-cj again, thank you very much, but i still have no idea how to prepare my dataset.
Could you please tell me how can I use the pokemon dataset or how to create the dataset for the training

> You need to use different config file for different dataset, model, image-size, multi-scale training, also you can create your own config file, like you said, if you want to train on the [pokemon](https://huggingface.co/datasets/lambdalabs/pokemon-blip-captions) dataset.

Very confusing.
6.1 What config file? How can I make it?

**Reply by kopyl** [reply]
10-11.1

How to prepare them in advance?

10-11.2

> Otherwise, in one iteration, you need to first extract the text feature using T5 and img feature using VAE, which is time consuming and occupy a lot of GPU memory.

Do I need to write the code to extract it? Or it's done automatically?

**Reply by kopyl** [reply]
10-11.3

> which is time consuming and occupy a lot of GPU memory.

How much memory?

**Reply by kopyl** [reply]
> you can look through the codebase in this repo

I did. Still nothing is clear :(

**Reply by lawrence-cj** [reply]
5.1: multi-scale input means the image size during training is not Square (i.e. 1024x1024), it may be 1024x512 which is according to the real image size. We don't use crop and resize in the multi-scale training. For more details: '2.3 Multi-Aspect Training' in https://arxiv.org/pdf/2307.01952.pdf
6.1 The config files talking about are in https://github.com/PixArt-alpha/PixArt-alpha/tree/master/configs/pixart_config, they are all examples for you to contruct your own config files changing the content (eg. [data](https://github.com/PixArt-alpha/PixArt-alpha/blob/fe0cb78065d64c18ecd8955a04e4f29138d47946/configs/pixart_config/PixArt_xl2_img1024_internalms.py#L5) and so on) according to your dataset. I assume you are pretty good at python and Diffusion models. I can't write the exact config file and training code for you, sry for that.
10-11: [This file](https://github.com/PixArt-alpha/PixArt-alpha/blob/master/tools/extract_features.py) is what you need to extract features in advance. 
I encourage you to study about the GPU memory consumption by yourself, in that way you may familiar with the codebase more.

**Reply by kopyl** [reply]
6.2 Well, why didn't you follow the most common format for you initial training, which is Huggiface's image datasets format?

**Reply by kopyl** [reply]
6.1.1 I'm not asking to give me the config file, i just want to understand how should i need to format a dataset for Pixart full fine-tune...

>  I can't write ... training code

Well, isn't the code already on this repo?

**Reply by lawrence-cj** [reply]
The training code exists. The reason why we don't follow the HF's image dataset format is at the begining of the project, not everything is settled and changing the dataset format is not the most important work now. Maybe we will change it in the future, or if you are intersted in our project, we highly recomment you to change it and push a PR for our code base. Appreciate it.

**Comment by kopyl** [comment]
> maybe using diffusers training pipeline instead is a better choice

Could you please tell how to do it?

**Reply by lawrence-cj** [reply]
Training pipeline in diffusers in not our duty, please refer to diffusers lib for more details. lol

**Reply by kopyl** [reply]
:(

**Comment by kopyl** [comment]
4.1

> [here ](https://github.com/PixArt-alpha/PixArt-alpha/blob/master/diffusion/data/datasets/InternalData_ms.py)have several dataset class for you to construct your own

Where? I can only see one dataset class `InternalDataMS`

**Reply by lawrence-cj** [reply]
https://github.com/PixArt-alpha/PixArt-alpha/tree/master/diffusion/data/datasets

**Comment by kopyl** [comment]
7.1

> the information in the txt file is only the image path like 'dir/sa_xxxxx.jpg'

Could you please show how it looks? I would love to see how I need to format it

**Reply by lawrence-cj** [reply]
I just copy the content in my txt file, looks like:
sa_3631
sa_3196
sa_2071
sa_886
sa_7088
sa_353
sa_1164
sa_2862
sa_2410
sa_732
sa_3250

and they are all SAM image names.

**Reply by kopyl** [reply]
Shouldn't images have extensions?

**Reply by kopyl** [reply]
Shouldn't part0.txt file have captions instead of file names?

**Reply by kopyl** [reply]
Also why do you have partitions for captions and don't have for images? So counterintuitive.

Do you have to name them following some format?

**Reply by lawrence-cj** [reply]
Here is why we do not need a extension: https://github.com/PixArt-alpha/PixArt-alpha/blob/master/diffusion/data/datasets/SA.py#L42C26-L42C26.
we store images and captions with the same name and different extensions, eg. imagedir/sa_3631.jpg and textdir/sa_3631.txt, that's why we don't need a extention in partition. the file in partition is convinient for us to use a subset of the SAM for usage.

**Reply by kopyl** [reply]
@lawrence-cj thank you very much

**Comment by kopyl** [comment]
15. I see this line in the config `model = 'PixArt_XL_2'`
But what is it? On GitHub i only see these models:
- T5
- VAE
- PixArt-α-SAM-256
- PixArt-α-512
- PixArt-α-1024

**Reply by lawrence-cj** [reply]
You can clone the whole codebase and search for `PixArt_XL_2`, which is located at:
https://github.com/PixArt-alpha/PixArt-alpha/blob/fe0cb78065d64c18ecd8955a04e4f29138d47946/diffusion/model/nets/PixArt.py#L282

**Reply by kopyl** [reply]
Are you talking about training from scratch?

**Reply by lawrence-cj** [reply]
Why say that? Maybe we can discuss more in Discord channel.

**Reply by kopyl** [reply]
@lawrence-cj just sent the message on the Discord channel :)

**Comment by kopyl** [comment]
16. Do I have to download the model or it's done automatically?

**Reply by lawrence-cj** [reply]
It should be downloaded automatically. refer: https://github.com/PixArt-alpha/PixArt-alpha/blob/master/tools/download.py
The codebase is kind of robust and growing, that's why I suggest you guys to look through the whole codebase for better future use.
